### PR TITLE
[DI] Enhance service locator error message

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #24555
| License       | MIT
| Doc PR        | -

See linked issue.

### CODE REVIEW

**Review by chalasr — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php [hunk 1]
```diff
@@ -78,6 +78,9 @@ protected function processValue($value, $isRoot = false)
     public static function register(ContainerBuilder $container, array $refMap)
     {
         foreach ($refMap as $id => $ref) {
+            if (!$ref instanceof Reference) {
+                throw new InvalidArgumentException(sprintf('Invalid service locator definition: only services can be referenced, "%s" found for key "%s". Inject parameter values using constructors instead.', is_object($ref) ? get_class($ref) : gettype($ref), $id));
```
Comment (yceruto) [inline src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php]: Please, see https://github.com/symfony/symfony/issues/24555#issuecomment-337071029 before merging this.

**Comment by stof** [comment]
The advice to inject the param in the constructor in ServiceLocatorTagPass may be weird, as this may happen for other cases than controller arguments. The error specific to controller arguments may be easier to throw in the RegisterControllerArgumentLocatorsPass

**Comment by nicolas-grekas** [comment]
I agree that the controller argument locator needs to deal with non-services before this is hit.
Yet, I'd still propose to merge this, as parameters should always be injected via the constructor, so the hint is 99% of the time correct, isn't it?

**Comment by yceruto** [comment]
I think throwing this exception still feels unexpected. e.g. let's suppose we have this configuration:
```yaml
services:
    _defaults:
        # ...
        bind:
            $foo: '%foobar%'
```
and `$foo` is binded correctly to any controller constructor, but in another controller I have this edit action (nothing to do with the intention of bind such a parameter):
```php
/**
 * @Route("/{$foo}/edit")
 */
public function editAction(string $foo) {}
```
triggering:
> Invalid service locator definition: only services can be referenced, "string" found for key "foo". Inject parameter values using constructors instead. 

oO! we won't know when it is really intentional or not.

**Comment by nicolas-grekas** [comment]
@yceruto I agree, PR welcomed, but unrelated to this one now :)

**Review by chalasr — APPROVED** [review]
👍

**Pull request merged**